[clang][modules-driver] Add scanner to detect C++20 module presence

naveen-seth opened the pull request:
This PR is part of a series to natively support C++20 module usage from the Clang driver (without requiring an external build system).
This introduces a new scanner that detects C++20 module usage in source files without using the preprocessor or lexer.

For now, it is enabled only with the `-fmodules-driver` flag and serves solely diagnostic purposes. In the future, the scanner will be enabled for any (modules-driver compatible) compilation with two or more inputs, and will help the driver determine whether to implicitly enable the modules driver.

Since the scanner adds very little overhead, we are also exploring enabling it for compilations with only a single input. This approach could allow us to detect `import std` usage in a single-file compilation, which would then activate the modules driver.
For performance measurements on this, see
https://github.com/naveen-seth/llvm-dev-cxx-modules-check-benchmark.

RFC:
https://discourse.llvm.org/t/rfc-modules-support-simple-c-20-modules-use-from-the-clang-driver-without-a-build-system

naveen-seth committed:
[clang][modules-driver] Add scanner to detect C++20 module presence

This PR is part of a series to natively support C++20 module usage
from the Clang driver (without requiring an external build system).
This introduces a new scanner that detects C++20 module usage in
source files without using the preprocessor or lexer.

For now, it is enabled only with the `-fmodules-driver` flag and
serves solely diagnostic purposes. In the future, the scanner will be
enabled for any (modules-driver compatible) compilation with two or
more inputs, and will help the driver determine whether to implicitly
enable the modules driver.

Since the scanner adds very little overhead, we are also exploring
enabling it for compilations with only a single input.
This approach could allow us to detect `import std` usage in a
single-file compilation, which would then activate the modules driver.
For performance measurements on this, see
https://github.com/naveen-seth/llvm-dev-cxx-modules-check-benchmark.

RFC:
https://discourse.llvm.org/t/rfc-modules-support-simple-c-20-modules-use-from-the-clang-driver-without-a-build-system

naveen-seth committed:
Remove anonymous namespace

See https://llvm.org/docs/CodingStandards.html#restrict-visibility

ChuanqiXu9 reviewed (commented):
I prefer to change the term "explicit module" in the code to something like `CXX20 module` or `named module` or whatever. Since when I read the term `explicit module` I assumed it is a style to build clang header modules.

naveen-seth commented:
> This will also eventually support C++20 named modules with Clang modules acting as header units, as well as just stand alone Clang modules.

Because this will also support stand alone Clang modules, I think the naming shouldn't be specific to only standard C++ named modules.

I adopted the term "explicit module build" from the Swift project. They implemented a comparable feature in their re-implemented driver (written in Swift), which is also used for Clang modules and makes use of Clang's dependency scanning tooling.

They explain it further in this announcement post:
[Swift Forum:  Explicit Module Builds, the new Swift Driver, and SwiftPM](https://forums.swift.org/t/explicit-module-builds-the-new-swift-driver-and-swiftpm/36990#explicit-module-builds-3)
and here's (the core of) their implementation:
[swift-driver/Sources/SwiftDriver/ExplicitModuleBuilds](https://github.com/swiftlang/swift-driver/tree/def5462e9daefe31da6fc8c132885ece416b2a9f/Sources/SwiftDriver/ExplicitModuleBuilds)

As an alternative naming suggestion for this feature, I also propose "Module dependency aware compilation/builds". 
In the code, the corresponding changes could look like this:
  
```
def fmodules_driver : Flag<["-"], "fmodules-driver">,
  Group<f_Group>, Visibility<[ClangOption]>,
  HelpText<"Enable support for module dependency aware compilation/build planning (experimental)">;
  ```

`Driver::BuildModuleDependencyAwareActions`

naveen-seth committed:
Address review feedback from Bigcheese

naveen-seth committed:
Fix test failure

Because #145220 removes the remark which is always emitted if the
modules driver is enabled, some tests fail because Clang emits no
output at all. FileCheck will then complain about empty stdin.
We fix this by using '-v' to ensure any output.

Bigcheese commented on clang/test/Driver/modules-driver-cxx20-module-usage-scanner.cpp:
```diff
@@ -0,0 +1,171 @@
+// RUN: %clang -std=c++23 -ccc-print-phases -fmodules-driver -Rmodules-driver \
+// RUN: %t/multiline-directive.cpp %t/empty.cpp 2>&1 | FileCheck %s --check-prefix=CHECK10
+// CHECK10: remark: found C++20 module usage in file '{{.*}}' [-Rmodules-driver]
+#define MACRO(a,  \
+              b)  \
+        call((a), \
+             (b)
+module;
```
This case has the same problem as above. `import m;` works here too.

Bigcheese reviewed (approved):
With the last bit this looks fine to me, but you should wait for Chuanqi and see if the naming should be changed.

naveen-seth committed:
Address review feedback (invalid test per P1857R3)

ChuanqiXu9 commented on clang/include/clang/Basic/DiagnosticDriverKinds.td:
```diff
@@ -577,6 +577,16 @@ def err_drv_reduced_module_output_overrided : Warning<
+def remark_found_cxx20_module_usage : Remark<
+  "found C++20 module usage in file '%0'">,
+  InGroup<ModulesDriver>;
+def remark_performing_explicit_module_build : Remark<
```
I don't like the name "explicit module". It is confusing to me. Please change this to other names. e.g., "clang managed C++ modules build"

ChuanqiXu9 commented on clang/include/clang/Driver/Driver.h:
```diff
@@ -790,6 +792,35 @@ class Driver {
+  /// \param C - The compilation that is being built.
+  /// \param Args - The input arguments.
+  /// \param Actions - The list to store the resulting actions onto.
+  void BuildExplicitModuleBuildActions(Compilation &C,
```
ditto, please change the name `explicit module`.

ChuanqiXu9 reviewed (changes requested):
I still hope we can get a new name. It is important for me to understand if some code may affect C++20 module or not by its name. I feel this can be helpful for other developers too.

naveen-seth committed:
[NFC] Use `--allow-empty` instead of `clang -v` & formatting

naveen-seth committed:
Renaming / Address review feedback from ChuanqiXu9

naveen-seth commented:
Done. For the regression test, I've updated it to use `--allow-empty` instead of forcing output using `clang -v` (+formatting & removed duplicate flags). Thank you for the reviews.
